server/notebooks/views.py

```diff
@@ -216,3 +217,47 @@ def tryit_view(request):
             "eval_frame_origin": EVAL_FRAME_ORIGIN,
         },
     )
+
+
+@require_http_methods(["POST"])
+@csrf_exempt
+def from_template_view(request):
+    iomd = urllib.parse.unquote(request.POST.get("iomd", ""))
+    if not iomd:
+        return HttpResponseBadRequest("Must specify iomd template")
+
+    title = request.POST.get("title")
+
+    # NOTE: We can not sadly not use files (request.FILES) attached to forms, as
+    # it is not possible to attach those programatically from the client side via
+    # JavaScript
```

Comment:
I'm not sure if that is true, IIRC you can use the fetch API to do that, and it should show up on the Django side correctly. [StackOverflow comment talking about this](https://stackoverflow.com/a/51109645)

Reply:
It is possible to do this with fetch, but that won't work since we're doing this anonymously (and we don't want to persist anything on the server). 

What we need to do is use an HTML form and attach the files to it programatically using JavaScript, but my initial investigations seemed to indicate that was impossible. On more recent browsers it is possible to do this, but only if the user has already made files available through a user input action (so it doesn't help us so much):

https://stackoverflow.com/questions/47515232/how-to-set-file-input-value-when-dropping-file-on-page

I'll update the comment before landing.

Reply:
I would recommend using a technique like the Django FileField to do this since even k8s nodes have a `/tmp` directory to persist files temporarily. My worry is mostly around this writing an own file handling, which for a public and anonymous upload API is really a bad idea.

That said, maybe I'm missing how this API will be used, usually all kinds of CLIs like curl can send regular form files, and if there is an integration in a frontend scenario there would need to be a file selector anyway. Could you elaborate what the use case of the API is?

Reply:
There are a few use cases:

* Creating a notebook via a python script by opening a page and submitting a hidden form (see the demo repo here: https://github.com/wlach/create-iodide-notebook)
* Creating a notebook via a web extension (by creating a hidden form and submitting it)
* Creating a notebook from a web application (by creating a hidden form and submitting it)

There are no cases that I have in mind that would use a file selector directly, although it would be nice to allow it.

--

Doing even more research, it looks like what I thought impossible *is* possible, using a DataTransfer object! See this github issue and the associated fiddle:

https://github.com/whatwg/html/issues/3269#issue-278914102

I'll give this a try and get back to you later :)

Reply:
Woohoo! It worked 🎉 We're now using Django's file handling here.